### DISCUSSION TITLE
Self Labels Data and Training Set running YOLOv8 on Google Coral

### DISCUSSION THREAD

**Discussion by sangyo1** [ultralytics/ultralytics#8768] [Q&A] [answered]
Currently I am using ROBOFLOW to generate my labeled dataset. I am using that to Google Colab "How to Train YOLOv8 Object Detection on a Custom Dataset" However I try deploying this train dataset to Raspberry Pi4, however it detect an object inference time is very slow. it is about 1000ms. I am curious is it possible to running this model with Google Coral TPU. How fast it can be? is it possible to run in RT? currently run create the training set with yolov8n, but I want to use yolov8m or yolov8s if I can run this faster which will be less than 200ms.
Also, one more question about image size.
So my original image size was 1920x1080. I labeled the data and create dataset as 640x640. However, when I run 

    yolo task=segment mode=predict imgsz=640 model=~/yolo8v/best_seg.pt conf=0.25 source=~/yolo8v/yolo8v_test/test/images

The model size become per image at shape (1, 3, 480, 640). how should I create right image size?
Also, what would be the best image size for training? I want reduce the image size up to 320x240, but I am worry about that will decrease the accuracy or something.

**Comment by glenn-jocher** [comment, marked as answer]
Hey there! 🌟 It's great to see your enthusiasm for optimizing YOLOv8 performance on edge devices like the Raspberry Pi4 and potentially Google Coral TPU. Let's address your queries:

1. **Running YOLOv8 on Google Coral TPU**: While YOLOv8 models are optimized for speed and accuracy, deploying them on Google Coral requires converting the model to a TPU-compatible format (e.g., TensorFlow Lite). The inference speed can indeed be faster on Coral due to its hardware acceleration, potentially reaching your target of less than 200ms. However, the exact speed gain depends on the model complexity and optimizations. For YOLOv8, you might start with `yolov8n` or `yolov8s` for a balance between speed and accuracy.

2. **Image Size for Training and Prediction**: When you notice the model input size changing to (1, 3, 480, 640), it's due to automatic resizing to maintain the aspect ratio of your original images. For maintaining a specific size like 640x640 without distortion, consider pre-processing your images to this size before training.

   As for reducing the training image size to 320x240, it's a trade-off. Smaller images will indeed speed up training and inference but might reduce the model's ability to detect small or detailed objects. It's worth experimenting to find the right balance for your specific use case.

Here's a quick tip for resizing images for training while maintaining aspect ratio:

```python
from PIL import Image

def resize_image(input_path, output_path, size=(640, 640)):
    image = Image.open(input_path)
    image = image.resize(size, Image.ANTIALIAS)
    image.save(output_path)
```

Remember, the best approach is to experiment with different model sizes (`yolov8n`, `yolov8m`, `yolov8s`) and image resolutions to find what works best for your specific scenario. Good luck! 🚀

**Reply by sangyo1** [reply]
Thank you so much for your answer. I have some questions for you.
Does accuracy decrease if I train the model with a different size? 
I know YOLOv8 has been pre-trained with a 640-pixel size. If I choose a different size, would it be bad for accuracy? 

I am also curious about using it for real-time processing with object detection. 
If I use Google Coral, how do I deploy the model? Would you recommend me to use YOLOv5 or YOLOv8? 
I want to minimize CPU usage on the Raspberry Pi.

**Reply by glenn-jocher** [reply]
@sangyo1 hello! Happy to help with your questions! 😊

- **Training with Different Sizes:** Changing the training size can affect accuracy, but not always negatively. YOLOv8 is quite adaptable to different image sizes. Smaller sizes train faster but might miss finer details, while larger sizes can capture more detail at the cost of speed. For real-time applications where you balance speed and accuracy, experimenting with different sizes is key.

- **Real-Time Processing with Google Coral:** For deploying on Google Coral, converting your model to TensorFlow Lite format optimized for the Edge TPU is essential. Both YOLOv5 and YOLOv8 are great, but YOLOv8 offers newer architecture improvements. If minimizing CPU usage on the Raspberry Pi is crucial, YOLOv8 might give better performance due to its efficiency. Remember, the final choice should depend on your specific use case performance after trying both.

Here's a quick snippet on how you might convert a PyTorch model (like YOLOv8) to TensorFlow Lite for Coral:

```python
# Assume you've already exported your YOLOv8 model to ONNX format
# Let's convert ONNX to TensorFlow SavedModel first
!onnx-tf convert -t tf -i "your_model.onnx" -o "tf_model"

# Then, convert the SavedModel to TensorFlow Lite
import tensorflow as tf

converter = tf.lite.TFLiteConverter.from_saved_model("tf_model")
converter.optimizations = [tf.lite.Optimize.DEFAULT]
tflite_model = converter.convert()

# Save the TFLite model
with open('model.tflite', 'wb') as f_out:
    f_out.write(tflite_model)
```

This example assumes you're comfortable converting models between formats. The specific commands and methods might need adjustments based on your model's details and environment setup.

Good luck with your project, and feel free to share your experiences or further questions! 🚀
